[shard] add ShardedOptimizer

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69569
* __->__ #68607

This PR added ShardedOptimizer and a API to get module parameters along with ShardedTensor param, it allows user to use this Optimizer Wrapper to construct a optimizer that involves ShardedTensor

The state_dict support will be a follow up diff

Differential Revision: [D32539994](https://our.internmc.facebook.com/intern/diff/D32539994/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32539994/)!

Commit:
[shard] add ShardedOptimizer

This PR added ShardedOptimizer and a API to get module parameters along with ShardedTensor param, it allows user to use this Optimizer Wrapper to construct a optimizer that involves ShardedTensor

The state_dict support will be a follow up diff

Differential Revision: [D32539994](https://our.internmc.facebook.com/intern/diff/D32539994/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32539994/)!

[ghstack-poisoned]

pritamdamania87 commented:
Looks like there is a CI failure

pritamdamania87 commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -0,0 +1,61 @@
+    ShardedTensor
+)
+
+def module_named_params_with_sharded_tensor(
```
nit: Maybe just call this `named_params_with_sharded_tensor`

pritamdamania87 commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -0,0 +1,61 @@
+    module: nn.Module,
+    prefix: str = '',
+    recurse: bool = True,
+) -> Iterator[Tuple[str, Union[nn.Parameter, ShardedTensor]]]:
```
Why have a `Tuple` here instead of just `Mapping`?

pritamdamania87 commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -0,0 +1,61 @@
+
+    r"""Returns an iterator over module parameters (together with the
+    ShardedTensor parameters), yielding both the name of the parameter
+    as well as the parameter itself. This is typically passed to an
```
nit: a ShardedOptimizer

pritamdamania87 commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -0,0 +1,61 @@
+    r"""Returns an iterator over module parameters (together with the
+    ShardedTensor parameters), yielding both the name of the parameter
+    as well as the parameter itself. This is typically passed to an
+    ShardedOptimizer
```
You probably want to use :class:`torch.distributed._sharded_optim.ShardedOptimizer` or just ``ShardedOptimizer`` here.

pritamdamania87 commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -0,0 +1,61 @@
+        (string, Union[Parameter, ShardedTensor]): Tuple containing
+            the name and parameter (or ShardedTensor parameter)
+
+    Example::
+
+        >>> for name, param in self.named_parameters():
+        >>>    if name in ['bias']:
+        >>>        print(param.size())
+
+    """
```
This example doesn't seem correct and seems like is just copy-paste from `named_parameters`.

pritamdamania87 commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -0,0 +1,61 @@
+    for mod_prefix, mod in modules:
+        # find all sharded tensor params
+        for name, val in vars(mod).items():
+            if name.startswith("__"):
+                continue
+
+            if val is None:
+                continue
+
```
Do we need this 2 statements? Shouldn't `if isinstance(val, ShardedTensor)` be sufficient?

pritamdamania87 commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -0,0 +1,61 @@
+                continue
+
+            if isinstance(val, ShardedTensor):
+                memo.add(val)
```
Why don't we check for whether val is already in `memo` here?

pritamdamania87 replied:
Need to address the comment above?

pritamdamania87 commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -0,0 +1,61 @@
+
+        # find all nn.Parameters
+        for name, val in mod._parameters.items():
+            if val is None or val in memo:
+                continue
+            memo.add(val)
```
It looks like this `memo` logic is similar to what is done in `named_parameters`. Does this mean if two parameters point to the same Tensor, only one of the parameters it returned by `named_parameters`? If so, isn't this confusing where not all parameters are returned?

pritamdamania87 commented on torch/distributed/_sharded_optim/api.py:
```diff
@@ -0,0 +1,61 @@
+from torch.distributed._sharded_tensor import ShardedTensor
+
+
+class ShardedOptimizer(object):
```
Shouldn't this be a subclass of `torch.optim.Optimizer`?

pritamdamania87 commented on torch/distributed/_sharded_optim/api.py:
```diff
@@ -0,0 +1,61 @@
+        self.param_groups = self._optim.param_groups
+        self.state = self._optim.state
+
+    def zero_grad(self, set_to_none: bool = False):
+        self._optim.zero_grad(set_to_none)
+
+    def step(self, closure=None):
+        """
+        Regular optimizer step
+        """
+        self._optim.step(closure)
```
Probably just copy the docs from `torch.optim` for these two methods.

pritamdamania87 commented on torch/distributed/_sharded_optim/api.py:
```diff
@@ -0,0 +1,61 @@
+        raise NotImplementedError("ShardedOptimizer state_dict not implemented yet!")
+
+
+    def load_state_dict(self, state_dict: Mapping[str, Any]):
```
What about the `add_param_group` API? It is part of the `torch.optim.Optimizer` interface.

pritamdamania87 commented on test/distributed/_sharded_optim/test_sharded_optim.py:
```diff
@@ -0,0 +1,94 @@
+        # optimizer update
+        optimizer.step()
+
+        # make sure the parameters (including sharded param)
+        # get updated by the optimizer
+        for key, val in before_update.items():
+            new_val = optimizer.params[key]
+            if isinstance(val, sharded_tensor.ShardedTensor):
+                self.assertNotEqual(
+                    val.local_shards()[0].tensor,
+                    new_val.local_shards()[0].tensor
+                )
+            else:
+                self.assertNotEqual(val, new_val)
+
```
To be more precise, can we also initialize a local optimizer directly and compare that the results of local optimizer and sharded match?

pritamdamania87 commented on test/distributed/_sharded_optim/test_sharded_optim.py:
```diff
@@ -0,0 +1,94 @@
+# Owner(s): ["oncall: distributed"]
+
+import torch
```
Don't we need to add this file to `run_test.py`?

Commit:
Update on "[shard] add ShardedOptimizer"


This PR added ShardedOptimizer and a API to get module parameters along with ShardedTensor param, it allows user to use this Optimizer Wrapper to construct a optimizer that involves ShardedTensor

The state_dict support will be a follow up diff

Differential Revision: [D32539994](https://our.internmc.facebook.com/intern/diff/D32539994/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32539994/)!

[ghstack-poisoned]

pritamdamania87 commented on test/distributed/_sharded_optim/test_sharded_optim.py:
```diff
@@ -89,6 +99,13 @@ def test_sharded_optim(self):
             else:
                 self.assertNotEqual(val, new_val)
 
+        # run second forward with updated parameters
```
nit: Any reason to do a second forward instead of just comparing the local and sharded parameters are the same?

wanchaol replied:
no particular reason haha, just thought it was easier to do that, updated to compare the weights directly

pritamdamania87 commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -30,9 +30,11 @@ def module_named_params_with_sharded_tensor(
+        >>> shard_parameter(model, "weight", spec)
+        >>> for name, param in named_params_with_sharded_tensor(model):
+        >>>    if name in ['weight']:
+        >>>        print(param.sharding_spec())
```
nit: Can just do `param.size()` here similar to the `named_parameters` example?

pritamdamania87 commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -0,0 +1,57 @@
+                name = mod_prefix + ('.' if mod_prefix else '') + name
+                yield name, val
+
+        # find all nn.Parameters
```
Can we just call `module.named_parameters()` to get this list instead of rewriting some of that logic here? I feel if something changes in `named_parameters()` that won't be reflected here and we will have an inconsistency.

pritamdamania87 commented on torch/distributed/_sharded_optim/api.py:
```diff
@@ -0,0 +1,89 @@
+        Collects all Tensors and local shards of ShardedTensor, uses these
+        Tensors as ``params`` for the optimizer while building the optimizer
+        using ``optimizer_class``, ``*optimizer_args`` and
+        ``*optimizer_kwargs``.
```
nit: docs here probably need Args:

pritamdamania87 commented on torch/distributed/_sharded_optim/api.py:
```diff
@@ -0,0 +1,89 @@
+class ShardedOptimizer(optim.Optimizer):
+    def __init__(
+        self,
+        params: Mapping[str, Union[Tensor, ShardedTensor]],
```
nit: `named_params` might be more appropriate here.

pritamdamania87 commented on torch/distributed/_sharded_optim/api.py:
```diff
@@ -0,0 +1,89 @@
+    def add_param_group(self, param_group: Any):
+        r"""Add a new param group
+        """
+        # TODO: implement load_state_dict
```
nit: fix comment.

pritamdamania87 reviewed (commented):
Looks good overall, mostly a bunch of nits.

zzzwen commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -0,0 +1,57 @@
+            are direct members of this module.
+
+    Yields:
+        (string, Union[Parameter, ShardedTensor]): Tuple containing
```
nit, `Parameter` does not feel like a counterpart of `ShardedTensor`, should this be `Tensor` instead?

zzzwen commented on torch/distributed/_sharded_optim/api.py:
```diff
@@ -0,0 +1,61 @@
+    def state_dict(self) -> Dict[str, Any]:
+        """
+        Returned state and param_groups will contain parameter keys
+        instead of parameter indices in torch.Optimizer.
```
Do you think it makes sense to have a "tensor" only interface like `Dict[str, Tensor]` in `ShardedOptimizer`?
If the vanilla optimizer use primitives, we can do the wrapping here.

wanchaol replied:
@zzzwen do you mean the "tensor" only interface for `ShardedOptimizer.state_dict()`? or you are referring to sth else

zzzwen replied:
Yes, I am talking about `ShardedOptimizer.state_dict()`

wanchaol replied:
I see, yeah `state_dict` it might make sense to have a tensor only interface, it will be done in a follow up PR https://github.com/pytorch/pytorch/pull/68608, still thinking about the correct semantics, lets discuss more there.

pritamdamania87 replied:
@zzzwen Not sure I understood the proposal here. Are you proposing adding two state_dict methods, one that returns Dict[str, Any] and one that returns Dict[str, Tensor]?

Commit:
Update on "[shard] add ShardedOptimizer"


This PR added ShardedOptimizer and a API to get module parameters along with ShardedTensor param, it allows user to use this Optimizer Wrapper to construct a optimizer that involves ShardedTensor

The state_dict support will be a follow up diff

Differential Revision: [D32539994](https://our.internmc.facebook.com/intern/diff/D32539994/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32539994/)!

[ghstack-poisoned]

fduwjj commented on torch/distributed/_sharded_optim/__init__.py:
```diff
@@ -0,0 +1,53 @@
+    memo = set()
+    for mod_prefix, mod in modules:
+        # find all sharded tensor params
+        for name, val in vars(mod).items():
```
Just a n00b question for my learning purpose. Looks like this API takes in a sharded tensor. Does this mean ShardedTensor is a subclass of Module? If so, is this logic for sharded tensor? Thanks!

wanchaol replied:
hmmm sorry I didn't quite understand your question, could you elaborate more? The logic of this API is to iterate through all module attributes, and find if there's sharded tensor attributes, if there is, we added it to the return param list.

pritamdamania87 commented on torch/distributed/_sharded_optim/api.py:
```diff
@@ -0,0 +1,98 @@
+    def state_dict(self) -> Dict[str, Any]:
+        """
+        Returned state and param_groups will contain parameter keys
+        instead of parameter indices like torch.Optimizer.
```
nit: torch.optim.Optimizer

pritamdamania87 commented on torch/distributed/_sharded_optim/api.py:
```diff
@@ -8,26 +8,35 @@
+        Args:
+            named_params (Dict[str, Union[Tensor, ShardedTensor]]) : a Dict
+                of parameters, where key is the parameter key, value is either
+                nn.Parameter or ShardedTensor parameter. This usually used in
```
We should say Tensor here since type above is using Tensor

Commit:
Update on "[shard] add ShardedOptimizer"


This PR added ShardedOptimizer and a API to get module parameters along with ShardedTensor param, it allows user to use this Optimizer Wrapper to construct a optimizer that involves ShardedTensor

The state_dict support will be a follow up diff

Differential Revision: [D32539994](https://our.internmc.facebook.com/intern/diff/D32539994/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32539994/)!

[ghstack-poisoned]

Commit:
Update on "[shard] add ShardedOptimizer"


This PR added ShardedOptimizer and a API to get module parameters along with ShardedTensor param, it allows user to use this Optimizer Wrapper to construct a optimizer that involves ShardedTensor

The state_dict support will be a follow up diff

Differential Revision: [D32539994](https://our.internmc.facebook.com/intern/diff/D32539994/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32539994/)!

[ghstack-poisoned]
